### PR TITLE
nixos/qemu-vm: add option for named network interfaces

### PR DESCRIPTION
**Pull request by GrahamDennis**
Adds a new option to the virtualisation modules that enables specifying explicitly named network interfaces in QEMU VMs. The existing virtualisation.vlans is still supported for cases where the name of the network interface is irrelevant.

This feature is useful for creating VMs with network configurations that exactly match the network configuration of real-world devices for more accurate SITL testing.

Note: This is PR https://github.com/NixOS/nixpkgs/pull/178290 which was later reverted in https://github.com/NixOS/nixpkgs/pull/213519 because it was causing some hydra tests to fail: https://github.com/NixOS/nixpkgs/pull/178290#issuecomment-1407751385.

This PR includes a fix for the "sit" and "privacy" tests that failed after the PR was merged last time, and adds a new test to cover the situation where the MAC addresses in the QEMU VM contain a character in the range [a-f].

###### Description of changes

* Added virtualisation.interfaces option to qemu-vm.nix.
* Updated the default value of virtualisation.vlans to empty if virtualisation.interfaces is not empty. This avoids a scenario where a user specifies virtualisation.interfaces, but inadvertently still has a network interface from the default virtualisation.vlans.
* Updated build-vms.nix to create the new network interfaces using QEMU options.
* Updated build-vms.nix to rename network interfaces using udev rules.
* Updated testing-python.nix to use the union of the VLANs specified in virtualisation.vlans and virtualisation.interfaces when collecting the list of VLANs.
* Updated networking tests in nixos/tests/networking.nix to use the new option where applicable. This cut down on unnecessary overrides to remove IP addresses from network interfaces.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by RaitoBezarius** [comment]
Awesome, please tag me for such PRs if you don't want specialized reviewers to miss them. :)

**Comment by GrahamDennis** [comment]
Thanks @RaitoBezarius! I have rebased and squashed my commits amending the commit title (and PR title) to match the guidelines.

I'll also need to move my change to the release notes after 23.05 branches.

**Comment by RaitoBezarius** [comment]
> Am I right in understanding that hydra hasn't run the PR yet? It looks to have unrelated evaluation errors.

At the moment, I see `ofborg-eval` cloning the project, so not running it yet.

File: nixos/lib/testing/network.nix [hunk 1]
```diff
@@ -18,24 +18,41 @@ let
...
+          [ { address = "192.168.${toString i.vlan}.${toString config.virtualisation.test.nodeNumber}";
               prefixLength = 24;
             }];
         });
 
+      qemuOptions = lib.flatten (forEach interfacesNumbered ({ fst, snd }:
+        qemu-common.qemuNICFlags snd fst.vlan config.virtualisation.test.nodeNumber));
+      udevRules = forEach interfacesNumbered ({ fst, snd }:
+        # MAC Addresses for QEMU network devices are lowercase, and udev string comparison is case-sensitive.
+        "SUBSYSTEM==\"net\",ACTION==\"add\",ATTR{address}==\"${toLower(qemu-common.qemuNicMac fst.vlan config.virtualisation.test.nodeNumber)}\",NAME=\"${fst.name}\"");
```
Comment (tfc) [inline nixos/lib/testing/network.nix:46]: This is just a nitpick, feel free to ignore, but if using `'''strings'''` we would have less backslash-noise in this rule.

Comment (GrahamDennis) [reply nixos/lib/testing/network.nix]: done. I've also rebased on master after the 23.05 release branched.

**Comment by GrahamDennis** [comment]
I don't understand why the manual check is failing. Does anyone have ideas?

**Comment by RaitoBezarius** [comment]
Potentially because I fucked up the initialization of the manual :-)

**Comment by RaitoBezarius** [comment]
If you can solve the conflicts, it'd be awesome. :)

**Comment by GrahamDennis** [comment]
Done @RaitoBezarius!

**Comment by RaitoBezarius** [comment]
> Done @RaitoBezarius!

I'm evaluating a recent nixos-unstable-small and we will evaluate this PR over https://hydra.nixos.org/jobset/nixos/python-test-refactoring to smoke out any regression. :)

**Comment by RaitoBezarius** [comment]
You should see one of your commit here: https://hydra.nixos.org/jobset/nixos/python-test-refactoring in ~30 mn. It will take some amount of time as the queue is currently full of other jobs and we don't have any priority.

Let's rediscuss when we will have some new (non flaky) failures.

**Comment by RaitoBezarius** [comment]
Please review the 4 new failures here: https://hydra.nixos.org/eval/1795393

I think aarch64 are classical spurious failures, please double check the x86_64 one.

If we are good, I think this is good to go.

**Comment by GrahamDennis** [comment]
None of those failures seem related to this change. The x86_64 change is a timeout, and it looks like control never passed from the kernel to userspace (https://hydra.nixos.org/build/221399319/log). I think we are good.
I wasn't aware that NixPkgs has flaky tests... this must be very frustrating

**Review by RaitoBezarius — COMMENTED** [review]
(was already reviewed in the previous PR)

**Pull request merged**